src/shims/unix/mem.rs

```diff
@@ -0,0 +1,244 @@
+//! This is an incomplete implementation of mmap/mremap/munmap which is restricted in order to be
+//! implementable on top of the existing memory system. The point of these function as-written is
+//! to allow memory allocators written entirely in Rust to be executed by Miri. This implementation
+//! does not support other uses of mmap such as file mappings.
+//!
+//! mmap/mremap/munmap behave a lot like alloc/realloc/dealloc, and for simple use they are exactly
+//! equivalent. But the memory-mapping API provides more control. For example:
+//!
+//! * It is possible to munmap a single page in the middle of a mapped region. We do not have a way
+//! to express non-contiguous allocations.
+//!
+//! * With MAP_FIXED it is possible to call mmap multiple times, but create a single contiguous
+//! range of mapped virtual addresses. A memory allocator can then choose to carve this up into
+//! allocations in arbitrary ways.
+//!
+//! For now, none of this is possible with the Miri implementation.
+
+use crate::*;
+use rustc_target::abi::{Align, Size};
+
+impl<'mir, 'tcx: 'mir> EvalContextExt<'mir, 'tcx> for crate::MiriInterpCx<'mir, 'tcx> {}
+pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriInterpCxExt<'mir, 'tcx> {
+    fn mmap(
+        &mut self,
+        addr: &OpTy<'tcx, Provenance>,
+        length: &OpTy<'tcx, Provenance>,
+        prot: &OpTy<'tcx, Provenance>,
+        flags: &OpTy<'tcx, Provenance>,
+        fd: &OpTy<'tcx, Provenance>,
+        offset: &OpTy<'tcx, Provenance>,
+    ) -> InterpResult<'tcx, Scalar<Provenance>> {
+        let this = self.eval_context_mut();
+
+        // We do not support MAP_FIXED, so the addr argument is always ignored
+        let addr = this.read_pointer(addr)?;
+        let length = this.read_scalar(length)?.to_target_usize(this)?;
```

Comment:
```suggestion
        let length = this.read_target_usize(length)?;
```
just while we are at it